File: flexlay/gui/minimap_widget.py
# Flexlay - A Generic 2D Game Editor
# Copyright (C) 2014 Ingo Ruhnke <grumbel@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


from typing import Optional

from PyQt5.QtCore import QSize
from PyQt5.QtGui import QPainter, QPaintEvent, QMouseEvent
from PyQt5.QtWidgets import QWidget

from flexlay.color import Color
from flexlay.editor_map import EditorMap
from flexlay.graphic_context import GraphicContext
from flexlay.gui.editor_map_component import EditorMapComponent
from flexlay.math import Point, Size, Rect, Rectf, Pointf, Sizef
from flexlay.sprite import Sprite
from flexlay.tool_context import ToolContext


class MinimapWidget(QWidget):

    def __init__(self, editormap_component: EditorMapComponent) -> None:
        super().__init__()

        self.drag_active = False
        self.last_serial = -1
        self.editor_map: Optional[EditorMap] = None
        self._parent = editormap_component
        self.minimap_surface: Optional[Sprite] = None

    def minimumSizeHint(self) -> QSize:
        return QSize(0, 0)

    def sizeHint(self) -> QSize:
        return QSize(300, 150)

    def paintEvent(self, event: QPaintEvent) -> None:
        if not self._parent.get_workspace().get_map():
            return

        if not self._parent or not self._parent.get_workspace():
            return

        painter = QPainter(self)
        gc = GraphicContext(painter)

        gc.push_modelview()

        # FIXME: Do this only on map changes
        if self.last_serial != self._parent.get_workspace().get_map().get_serial():
            # or editor_map != parent.get_workspace().get_map())
            self.update_minimap()
            self.last_serial = self._parent.get_workspace().get_map().get_serial()
            self.editor_map = self._parent.get_workspace().get_map()

        if True:
            # Draw background color
            gc.fill_rect(Rectf(0, 0, self.width(), self.height()),
                         Color(200, 200, 200, 225))

        # FIXME: This doesn't work all that well
        assert ToolContext.current is not None
        tilemap = ToolContext.current.tilemap_layer
        if tilemap and tilemap.height != 0 and tilemap.width != 0:
            tile_size = tilemap.get_tileset().get_tile_size()

            map_width = tilemap.width * tile_size
            map_height = tilemap.height * tile_size

            small_tile_size = Size(tile_size * self.width() // map_width + 1,
                                   tile_size * self.height() // map_height + 1)

            field = tilemap.field

            # FIXME: No current tileset
            if False:
                for y in range(0, field.height):  # type: ignore[unreachable]
                    for x in range(0, field.width):
                        tile = tilemap.get_tileset().create(field.at(x, y))
                        if tile:
                            gc.fill_rect(Rect.from_ps(Point((x * tile_size) * self.width() / map_width,
                                                            (y * tile_size) * self.height() / map_height),
                                                      small_tile_size),
                                         tile.get_color())

            if self.minimap_surface:
                self.minimap_surface.draw(0, 0, gc)
                # self.minimap_surface.draw(Rect.from_ps(Point(0, 0), Size(self.width(), self.height())))

            # Draw cursor
            rect = self._parent.get_clip_rect()
            screen_rect = Rectf.from_ps(Pointf(rect.left * self.width() / map_width,
                                               rect.top * self.height() / map_height),
                                        Sizef(rect.width * self.width() / map_width,
                                              rect.height * self.height() / map_height))
            gc.fill_rect(screen_rect, Color(255, 255, 0, 50))
            gc.draw_rect(screen_rect, Color(0, 0, 0))

        gc.pop_modelview()

    def update_minimap(self) -> None:
        # FIXME: This doesn't work all that well
        # tilemap = TilemapLayer.current
        pass
        # GRUMBEL
        # if tilemap:
        #     field = tilemap.field
        #     buffer = PixelBuffer(tilemap.width, tilemap.height)

        #     map_width = tilemap.width
        #     map_height = tilemap.height

        # FIXME: No Tileset.current
        #     buf = buffer.get_data()
        #     for y in range(0, map_height):
        #         for x in range(0, map_width):
        #             tile = tilemap.get_tileset().create(field.at(x, y))
        #             if tile:
        #                 buf[4 * (x + y * map_width) + 3] = tile.get_color().get_red()
        #                 buf[4 * (x + y * map_width) + 2] = tile.get_color().get_green()
        #                 buf[4 * (x + y * map_width) + 1] = tile.get_color().get_blue()
        #                 buf[4 * (x + y * map_width) + 0] = tile.get_color().get_alpha()
        #             else:
        #                 buf[4 * (x + y * map_width) + 0] = 0
        #                 buf[4 * (x + y * map_width) + 1] = 0
        #                 buf[4 * (x + y * map_width) + 2] = 0
        #                 buf[4 * (x + y * map_width) + 3] = 0
        #
        #    self.minimap_surface = Sprite(buffer)

    def mouseMoveEvent(self, event: QMouseEvent) -> None:
        # FIXME: This doesn't work all that well
        assert ToolContext.current is not None
        tilemap = ToolContext.current.tilemap_layer
        if tilemap:
            tile_size = tilemap.get_tileset().get_tile_size()
            map_width = tilemap.width * tile_size
            map_height = tilemap.height * tile_size

            if self.drag_active:
                self._parent.move_to(event.x() * map_width / self.width(),
                                     event.y() * map_height / self.height())
        self.repaint()

    def mousePressEvent(self, event: QMouseEvent) -> None:
        # FIXME: This doesn't work all that well
        assert ToolContext.current is not None
        tilemap = ToolContext.current.tilemap_layer
        if tilemap:
            tile_size = tilemap.get_tileset().get_tile_size()
            map_width = tilemap.width * tile_size
            map_height = tilemap.height * tile_size

            self._parent.move_to(event.x() * map_width / self.width(),
                                 event.y() * map_height / self.height())
            self.drag_active = True
            self.grabMouse()

        self.repaint()

    def mouseReleaseEvent(self, event: QMouseEvent) -> None:
        assert ToolContext.current is not None

        tilemap = ToolContext.current.tilemap_layer
        if tilemap:
            self.drag_active = False
            self.releaseMouse()
        self.repaint()


# EOF #

File: libflexlay/gui/minimap_widget.cpp
// Flexlay - A Generic 2D Game Editor
// Copyright (C) 2002 Ingo Ruhnke <grumbel@gmail.com>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

#include "gui/minimap_widget.hpp"

#include <QPainter>
#include <QMouseEvent>
#include <iostream>

#include "display.hpp"
#include "editor_map.hpp"
#include "editor_map_component.hpp"
#include "workspace.hpp"
#include "pixel_buffer.hpp"
#include "tile.hpp"
#include "tilemap_layer.hpp"
#include "tileset.hpp"
#include "graphic_context.hpp"

MinimapWidget::MinimapWidget(EditorMapComponent* p) :
  m_drag_active(false),
  m_last_serial(-1),
  m_editor_map(false),
  m_parent(),
  m_minimap_surface()
{
  m_parent = p;
}

QSize
MinimapWidget::minimumSizeHint() const
{
  return QSize(0, 0);
}
  
QSize
MinimapWidget::sizeHint() const
{
  return QSize(300, 150);
}

void
MinimapWidget::paintEvent(QPaintEvent* event)
{
  if (!m_parent->get_workspace().get_map()) 
    return;

  if (!m_parent || m_parent->get_workspace().is_null())
    return;
  
  QPainter painter(this);
  GraphicContext gc(painter);

  gc.push_modelview();

  // FIXME: Do this only on map changes
  if (m_last_serial != m_parent->get_workspace().get_map().get_serial())
    // || editor_map != parent->get_workspace().get_map())
  {
    update_minimap();
    m_last_serial = m_parent->get_workspace().get_map().get_serial();
    m_editor_map  = m_parent->get_workspace().get_map();
  }

  if (1)
  { // Draw background color
    gc.fill_rect(Rect(Point(0, 0), Size(width(), height())),
                 Color(200, 200, 200, 225));
  }

  // FIXME: This doesn't work all that well
  TilemapLayer tilemap = TilemapLayer::current();

  if (!tilemap.is_null() && tilemap.get_height() != 0 && tilemap.get_width() != 0)
  {
    int tile_size = tilemap.get_tileset().get_tile_size();

    int map_width  = tilemap.get_width()  * tile_size;
    int map_height = tilemap.get_height() * tile_size;

    Size small_tile(tile_size * width() / map_width + 1,
                    tile_size * height() / map_height + 1);

    Field<int>* field = tilemap.get_field();

    // FIXME: No current tileset
    if (0)
    {
      for(int y = 0; y < field->get_height(); ++y)
        for(int x = 0; x < field->get_width(); ++x)
        {
          Tile* tile = tilemap.get_tileset().create(field->at(x, y));
          if (tile)
            gc.fill_rect(Rect(Point((x * tile_size) * width() / map_width,
                                    (y * tile_size) * height() / map_height),
                              Size(small_tile)),
                         tile->get_color());
          gc.flush();
        }
    }
#ifdef GRUMBEL
    m_minimap_surface.draw(Rect(Point(0, 0),
                                Size(width(), height())));
#else
    m_minimap_surface.draw(0, 0, gc);
#endif

    // Draw cursor
    Rect rect(m_parent->get_clip_rect());
    Rect screen_rect(Point(rect.left * width()  / map_width,
                           rect.top * height() / map_height),
                     Size(rect.get_width() * width() /map_width,
                          rect.get_height()* height()/map_height));
    gc.fill_rect(screen_rect,
                 Color(255, 255, 0, 50));
    gc.draw_rect(screen_rect,
                 Color(0, 0, 0));
  }

  gc.pop_modelview();
}

void
MinimapWidget::update_minimap()
{
  // FIXME: This doesn't work all that well
  TilemapLayer tilemap = TilemapLayer::current();

  if (!tilemap.is_null())
  {
    Field<int>* field = tilemap.get_field();

    PixelBuffer buffer(tilemap.get_width(), tilemap.get_height());

    int map_width  = tilemap.get_width();
    int map_height = tilemap.get_height();

    // FIXME: No Tileset::current()
    unsigned char* buf = static_cast<unsigned char*>(buffer.get_data());
    for(int y = 0; y < map_height; ++y)
      for(int x = 0; x < map_width; ++x)
      {
        Tile* tile = tilemap.get_tileset().create(field->at(x, y));
        if (tile)
        {
          buf[4*(x + y * map_width) + 3] = tile->get_color().get_red();
          buf[4*(x + y * map_width) + 2] = tile->get_color().get_green();
          buf[4*(x + y * map_width) + 1] = tile->get_color().get_blue();
          buf[4*(x + y * map_width) + 0] = tile->get_color().get_alpha();
        }
        else
        {
          buf[4*(x + y * map_width) + 0] = 0;
          buf[4*(x + y * map_width) + 1] = 0;
          buf[4*(x + y * map_width) + 2] = 0;
          buf[4*(x + y * map_width) + 3] = 0;
        }
      }

    m_minimap_surface = Sprite(buffer);
  }
}

void
MinimapWidget::mouseMoveEvent(QMouseEvent* event)
{
  // FIXME: This doesn't work all that well
  TilemapLayer tilemap = TilemapLayer::current();
  if (!tilemap.is_null())
  {
    int tile_size  = tilemap.get_tileset().get_tile_size();
    int map_width  = tilemap.get_width()  * tile_size;
    int map_height = tilemap.get_height() * tile_size;

    if (m_drag_active)
      m_parent->move_to(event->x() * map_width / width(),
                        event->y() * map_height / height());
  }

  repaint();
}

void
MinimapWidget::mousePressEvent(QMouseEvent* event)
{
  // FIXME: This doesn't work all that well
  TilemapLayer tilemap = TilemapLayer::current();
  if (!tilemap.is_null())
  {
    int tile_size  = tilemap.get_tileset().get_tile_size();
    int map_width  = tilemap.get_width()  * tile_size;
    int map_height = tilemap.get_height() * tile_size;

    m_parent->move_to(event->x() * map_width / width(),
                      event->y() * map_height / height());
    m_drag_active = true;
    grabMouse();
  }

  repaint();
}

void
MinimapWidget::mouseReleaseEvent(QMouseEvent* event)
{
  TilemapLayer tilemap = TilemapLayer::current();
  if (!tilemap.is_null())
  {
    m_drag_active = false;
    releaseMouse();
  }

  repaint();
}

/* EOF */
